receiver/k8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,146 @@
+	client          k8s.Interface
+	logsConsumer    consumer.Logs
+	stopperChanList []chan struct{}
+	ctx             context.Context
+	cancel          context.CancelFunc
+	obsrecv         *obsreport.Receiver
+}
+
+// newReceiver creates the Kubernetes events receiver with the given configuration.
+func newReceiver(
+	set component.ReceiverCreateSettings,
+	config *Config,
+	consumer consumer.Logs,
+	client k8s.Interface,
+) (component.LogsReceiver, error) {
+	transport := "http"
+
+	return &k8seventsReceiver{
+		settings:     set,
+		config:       config,
+		client:       client,
+		logsConsumer: consumer,
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             config.ID(),
+			Transport:              transport,
+			ReceiverCreateSettings: set,
+		}),
+	}, nil
+}
+
+func (kr *k8seventsReceiver) Start(ctx context.Context, host component.Host) error {
+	kr.ctx, kr.cancel = context.WithCancel(ctx)
+
+	kr.settings.Logger.Info("Starting to watch namespaces for the events.")
+	if len(kr.config.Namespaces) == 0 {
+		kr.startWatch(corev1.NamespaceAll)
+	} else {
+		for _, ns := range kr.config.Namespaces {
+			kr.startWatch(ns)
+		}
+	}
+
+	return nil
+}
+
+func (kr *k8seventsReceiver) Shutdown(context.Context) error {
+	// Stop watching all the namespaces by closing all the stopper channels.
+	for _, stopperChan := range kr.stopperChanList {
+		close(stopperChan)
+	}
+	kr.cancel()
+	return nil
+}
+
+// Add the 'New Event' handler and trigger the watch for a specific namespace.
+func (kr *k8seventsReceiver) startWatch(ns string) {
+	stopperChan := make(chan struct{})
+	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
+	kr.startWatchingNamespace(kr.client, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
+	}, ns, stopperChan)
+}
+
+func (kr *k8seventsReceiver) handleNewEvent(ev *corev1.Event) {
+	if kr.allowEvent(ev) {
+		ld := k8sEventToLogData(kr.settings.Logger, ev)
+
+		ctx := kr.obsrecv.StartLogsOp(kr.ctx)
+		consumerErr := kr.logsConsumer.ConsumeLogs(ctx, ld)
+		kr.obsrecv.EndLogsOp(ctx, typeStr, 1, consumerErr)
+	}
+}
+
+// startWatchingNamespace creates an informer and starts
+// watching a specific namespace for the events.
+func (kr *k8seventsReceiver) startWatchingNamespace(
+	clientset k8s.Interface,
+	handlers cache.ResourceEventHandlerFuncs,
+	ns string,
+	stopper chan struct{},
+) {
+	client := clientset.CoreV1().RESTClient()
+	watchList := cache.NewListWatchFromClient(client, "events", ns, fields.Everything())
+	_, controller := cache.NewInformer(watchList, &corev1.Event{}, 0, handlers)
+	go controller.Run(stopper)
+}
+
+// Allow events with eventTimestamp(EventTime/LastTimestamp/FirstTimestamp)
+// not older than the current time by 'InitialLookback' so that
+// event flood can be avoided upon startup.
+func (kr *k8seventsReceiver) allowEvent(ev *corev1.Event) bool {
+	eventTimestamp := getEventTimestamp(ev)
```

Comment:
Is it safe to assume that after a certain period, all events are going to be allowed?

Reply:
Then perhaps have a flag somewhere where the calculation below can be skipped? You should save some allocations and processing time with that, which would benefit the collector during its whole lifecycle, except for the duration of `kr.config.InitialLookback`, which is likely very short in comparison.

Reply:
In the below [comment thread](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6885#discussion_r773709125), if we conclude to use `kr.startTime` instead of `time.now()`, then we can introduce a logic to skip this check after some time (5 min after receiver's `kr.startTime`).

Reply:
For the below two reasons we can't stop the filtering after some time:

1. We don't know how much time it'll take to process the backlog. If the user gives `InitialLookback` too big then the receiver might take more time to process the previous events (receiver processes only one event at a time). If we stop the filtering after `t` time and by that time if the processing of previous events is not completed by a receiver then the events older than `InitialLookback` (considering backlog has much older events than `InitialLookback` window) will also be allowed, which is not expected.
2. Events are not sorted with time so we can't have a logic of stop processing events once we receive the first event whose time is greater than the receiver start time.